File: brain/src/main/kotlin/plp/brain/TranscriptionService.kt
package plp.brain

import plp.logging.KotlinLogging
import kotlin.io.path.ExperimentalPathApi
import kotlin.io.path.createTempFile
import kotlin.io.path.deleteExisting
import kotlin.io.path.writeBytes
import java.nio.file.Path as NioFilePath

private val logger = KotlinLogging.logger {}

/**
 * Transcribe given audio
 *
 * @param modelPath path to wav2letter's model files
 * @param tmpDir location for temporary files. Can't use the system default because tmpfs isn't mounted correctly into Docker containers, producing empty transcriptions
 */
class TranscriptionService(private val modelPath: NioFilePath, private val tmpDir: NioFilePath) :
    TranscriptionServiceGrpcKt.TranscriptionServiceCoroutineImplBase() {
    @ExperimentalPathApi
    @Suppress("TooGenericExceptionCaught")
    override suspend fun transcribeFile(request: Transcription.TranscriptionRequest): Transcription.TranscriptionResponse {
        logger.info { "handling request ${request.id}" }

        return try {
            val tempFile = createTempFile(directory = tmpDir, suffix = ".wav")
            logger.debug { "storing request ${request.id} bytes in $tempFile" }
            tempFile.writeBytes(request.audio.toByteArray())

            val text = transcribeFile(tempFile.toFile(), modelPath.toFile())

            logger.debug { "cleaning up $tempFile now that transcription is done" }
            tempFile.deleteExisting()

            Transcription.TranscriptionResponse.newBuilder().setId(request.id).setText(text).build()
        } catch (e: Exception) {
            logger.error { e }
            throw e
        }
    }
}

File: transcribe_chrome/src/TranscriptionService.py
import tempfile
from pathlib import Path

import grpc
from sclog import getLogger

from plp.proto import Transcription_pb2
from plp.proto import Transcription_pb2_grpc
from audio2text import transcribe

logger = getLogger(__name__)

TRANSCRIPTION_SERVICE_PORT = 50057


def save_bytes_as_tmp_wav_file(b: bytes) -> str:
    """
    Save the given bytes to a file in a new temporary directory and return that file's path
    """
    fd, path = tempfile.mkstemp(suffix=".wav")
    with open(fd, "wb") as f:
        f.write(b)
    logger.debug("saved bytes to temporary file %s", path)
    return path


class TranscriptionService(Transcription_pb2_grpc.TranscriptionServiceServicer):
    def __init__(self, screenshot_dir: Path) -> None:
        super().__init__()
        self.screenshot_dir = screenshot_dir

    def TranscribeFile(self, request, context):
        logger.debug("received request %i", request.id)

        tmp_file = Path(save_bytes_as_tmp_wav_file(request.audio))
        # TODO Transcribe wave file at tmp_file here
        transcription = transcribe(
            str(tmp_file), screenshot_directory=self.screenshot_dir
        )
        tmp_file.unlink()

        return Transcription_pb2.TranscriptionResponse(
            id=request.id,
            text=transcription,  # TODO add text
        )